src/api/java/com/github/minecraftschurlimods/arsmagicalegacy/api/data/PrefabSpellProvider.java

```diff
@@ -0,0 +1,97 @@
+package com.github.minecraftschurlimods.arsmagicalegacy.api.data;
+
+import com.github.minecraftschurlimods.arsmagicalegacy.api.spell.IPrefabSpell;
+import com.google.gson.Gson;
+import com.google.gson.GsonBuilder;
+import com.google.gson.JsonObject;
+import net.minecraft.data.DataGenerator;
+import net.minecraft.data.DataProvider;
+import net.minecraft.data.HashCache;
+import net.minecraft.resources.ResourceLocation;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+
+import java.io.IOException;
+import java.util.HashMap;
+import java.util.Map;
+
+public abstract class PrefabSpellProvider implements DataProvider {
+    private static final Gson GSON = new GsonBuilder().setPrettyPrinting().create();
+    private static final Logger LOGGER = LogManager.getLogger();
+    private final DataGenerator generator;
+    private final String namespace;
+    private final Map<ResourceLocation, JsonObject> data = new HashMap<>();
+
+    public PrefabSpellProvider(String namespace, DataGenerator generator) {
+        this.namespace = namespace;
+        this.generator = generator;
+    }
+
+    protected abstract void createPrefabSpells();
+
+    @Override
+    public void run(HashCache pCache) throws IOException {
+        createPrefabSpells();
+        for (Map.Entry<ResourceLocation, JsonObject> entry : data.entrySet()) {
+            ResourceLocation id = entry.getKey();
+            JsonObject spell = entry.getValue();
+            DataProvider.save(GSON, pCache, spell, generator.getOutputFolder().resolve("data/" + id.getNamespace() + "/prefab_spells/" + id.getPath() + ".json"));
+        }
+    }
+
+    @Override
+    public String getName() {
+        return "Prefab Spells";
```

Comment:
The name should be different for each instance with a different output namespace